### ISSUE TITLE
Orchestration Stuck in Running State

### ISSUE THREAD

**Issue by jtoniolo** [Azure/azure-functions-durable-extension#904]
Orchestration (or Sub-Orchestration) gets stuck in Running State after Activity completion. Orchestration does not resume.

Where it stops seems completely random for us. We have a very long Orchestrator with many Activities and Sub-orchestrators and one fan-in/fan-out somewhere in the middle. Sometimes the main Orchestrator gets stuck in Running state, sometimes one of the Sub-orchestrators gets stuck.

App Insights doesn't appear to be logging any related exceptions.

Same issue with DF 1.8.3 and 1.7.1 (current version is 1.8.3)
Azure Functions v1

We first noticed this on 2019-08-16. Everything was running fine the previous weekend, from what I recall.

Account info:

Azure region: Canada East
Function app name: YMCAFunctionAppDEV
Function name: BulkCreate (orchestrator), BulkUpdate (orchestrator) or ProcessPaymentShares_ProcessBillDate (orchestrator)
Azure storage account name: ymcaappservicesstorage
Hub name: BDOTasksDEV20190826

Example of execution stuck in "Running" state:
Orchestration instance ID (partition key): 4a29999c78a9417f86529e76f3741743:2
Parent orchestration instance ID: c8cfb707ae254a969de553904a427896 (SINGLETON)
Time: 2019-08-27T14:24

Example of execution stuck in "Running" state:
Orchestration instance ID (partition key): 53f63730a9b7481ba3830e62e109bb5e:2
Parent orchestration instance ID: c8cfb707ae254a969de553904a427896 (SINGLETON)
Time: 2019-08-27T14:51 (failed around 15:12)

**Comment by jtoniolo** [comment]
The job seems to have resumed this morning around 1:30 am Eastern, just as mysteriously as the issue began. We're running a test (new instance) and will report back when it fails or succeeds.

**Comment by ConnorMcMahon** [comment]
I am taking a shot at this investigation now. Expect an update sometime this week.

**Comment by ConnorMcMahon** [comment]
@jtoniolo I think I understand what happened with 4a29999c78a9417f86529e76f3741743:2.

The last thing the orchestration did was schedule a suborchestration (b17bf7ca059f4cf8972038b071d6b387:419).  This occurred at 2019-08-27 14:26:21.4359392 UTC. However, the suborchestration didn't start up until 2019-09-03 05:43:31.5482588 UTC. Since the original orchestration was waiting on that suborchestration to finish executing, it didn't wake up until that day either.

As for why b17bf7ca059f4cf8972038b071d6b387:419 took so long to start executing, we see that when it finally did start, the dequeue count was 1911.  To us, this generally indicates that another Durable Function app was operating on the same storage account with the same taskhub.

**Comment by paulpce** [comment]
I think I have this same issue, except it happens every single time for me.  In spite of the last line of code in a sub-orchestration executing successfully, the instance remains in the Running state (as does its parent orchestration, since it's waiting for the sub-orchestration to complete), and the whole process just sits there doing nothing.  I haven't yet been able to make it succeed on the Azure-hosted site, although it works when running and debugging locally.

It's taken me _days_ to analyse this issue and I'm at my wits' end.

I've updated from DTFx v1.8.2 to v1.8.3 to no avail (#774).  I've tried different return types (POCO and JObject), I've cleared out the Instances and History tables (which were full of thousands of "Running" but incomplete instances), pored over logs and Instances/History tables, I even added new (unique GUID) instance ID's on orchestrator execution, instead of using the default of not supplying them (ref #746 and [this JS bug](https://github.com/Azure/azure-functions-durable-js/issues/54), thinking perhaps it was related) but I still can't see any reason why the sub-orchestration just stops running.

Region: Australia East
Function app: TDSQueueProcessorDevTest
Function names: QPOrchestrator (main), GetTxnAndFilterFromQDIOrchestrator (sub-orch)
Azure storage account name: devtesttdsstoragepce
QPOrchestrator instance ID: 11cd56eab9d14ad4a81f29e906cf7c13
QPOrchestrator execution ID: a1555aaf1a0a4be097e59f4939259d6b
GetTxnAndFilterFromQDIOrchestrator instance ID: 90bab913-e471-4ff9-ae91-50c2a6e6da94
GetTxnAndFilterFromQDIOrchestrator execution ID: 4f9a47fd9b284a52aa10ee5bc12f9959

Happy to add a new bug report if it's determined to be a different issue.

**Comment by paulpce** [comment]
@ConnorMcMahon any progress on this issue?  This is a blocker for us.  I'm happy to open a new issue if you think mine is different but it seems similar

**Comment by ConnorMcMahon** [comment]
@pualpce, Sorry for the delay, missed your previous update in the email. I will investigate your specific issue today.

**Comment by ConnorMcMahon** [comment]
@pualpce, I can definitely say that your issue is distinct from the original issue in this thread.

Looking at the logs, it looks like for some reason your suborchestration GetTxnAndFilterFromQDIOrchestrator is never successfully finishing, which is why your orchestrations are getting stuck. This is despite the fact that the scheduled activity functions all seem to be completing as expected.

Looking at our logs, nothing obviously wrong stands out to me. That being said, if EVERY execution is getting stuck, that doesn't indicate some strange race condition, but likely some violation of one of the durable orchestration code constraints. See [here](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-code-constraints) for more information.

In order to diagnose any further from our end, I would likely need to see some code. We could arrange to have you email it to us for a more private way to share your code with us.

**Comment by paulpce** [comment]
Thanks for looking into this @ConnorMcMahon, that would be great. I can't see any violations of the orchestration constraints, it's a relatively straightforward orchestration.  Let me know if I should open another issue to track this and how you'd like to get in touch.

**Comment by ConnorMcMahon** [comment]
@paulpce , I am going to close this issue since yours seems to be a different cause/symptoms. Could you open a new issue with the details again?

Your options for sharing code:
1. Post the code in the issue.
2. Share it as a gist.
3. Email me directly at Connor.McMahon@microsoft.com
4. Open a support ticket with Microsoft and use those channels to securely share code.

Obviously 1 and 2 require posting code publicly, so that may not be an option for you.

**Comment by paulpce** [comment]
*\*facepalm\** There actually *was* a hidden async call in there to that was added in a moment of stupidity. Poring over the code once more in preparation to send it to you, I noticed it and on removing the line of code the orchestration now completes. Thanks and I'm so sorry for wasting your time @ConnorMcMahon, and sorry to the OP for hijacking your issue.

**Comment by ConnorMcMahon** [comment]
Glad to hear the issue is fixed @paulpce!
